### ISSUE TITLE
Refactor FastAPI into layered architecture and remove AWS deployment coupling

### ISSUE THREAD

**Issue by TumeloKonaite** [TumeloKonaite/Medical-Insurance-Cost#32]
# Refactor FastAPI into layered architecture and remove AWS deployment coupling

## Summary

Refactor the application into a clear layered architecture with:

- A lean FastAPI/API layer
- Pydantic request and response schemas
- A prediction service layer
- A repository layer for persisted model artifacts
- Explicit dependency injection
- No AWS, S3, CodeBuild, Elastic Beanstalk, or `.ebextensions` configuration

The existing HTML and JSON prediction endpoints must remain functional.

## Current state

The application currently mixes several responsibilities:

- `main.py` contains:
  - FastAPI application setup
  - Routes
  - HTML rendering
  - Pydantic schemas
  - Prediction orchestration
  - Error handling
- `pipeline/predict_pipeline.py` contains:
  - Model and preprocessor paths
  - S3 download logic
  - Artifact loading
  - DataFrame construction
  - Prediction execution
- Training components persist artifacts directly through `src/utils.py`.
- Model artifacts are reloaded for each prediction.
- API tests train real models and monkeypatch S3-related implementation details.
- Pydantic is used indirectly through FastAPI but is not declared as a first-class dependency.

AWS-specific code and documentation currently exist in:

- `.ebextensions/healthcheck.config`
- `deploy/ha/`
- `buildspec-train.yml`
- `pipeline/predict_pipeline.py`
- `.env.example`
- `requirements.txt`
- `pyproject.toml`
- `uv.lock`
- `tests/test_fastapi_smoke.py`
- `README.md`
- `docs/issue-mvp-demo.md`

Current baseline:

- `5 passed`
- Ruff reports no violations

## Goal

Introduce this dependency direction:

```text
FastAPI routes
    |
    v
Pydantic schemas
    |
    v
PredictionService
    |
    v
ArtifactRepository interface
    |
    v
Local filesystem implementation
```

FastAPI-specific concepts must not leak into the service or repository layers.

## Proposed package structure

```text
src/
└── medical_insurance/
    ├── __init__.py
    ├── main.py
    ├── api/
    │   ├── __init__.py
    │   ├── dependencies.py
    │   └── routes/
    │       ├── __init__.py
    │       ├── health.py
    │       ├── predictions.py
    │       └── web.py
    ├── schemas/
    │   ├── __init__.py
    │   └── prediction.py
    ├── services/
    │   ├── __init__.py
    │   └── prediction_service.py
    ├── repositories/
    │   ├── __init__.py
    │   └── artifact_repository.py
    ├── training/
    │   ├── __init__.py
    │   ├── data_ingestion.py
    │   ├── data_transformation.py
    │   └── model_trainer.py
    └── exceptions.py
```

Exact filenames may vary, but the layer boundaries and dependency direction should remain explicit.

### 1. Create Pydantic schemas

Move API models out of `main.py`.

Create a shared `PredictionRequest` model containing:

- `age`: integer, greater than or equal to zero
- `sex`: `female` or `male`
- `bmi`: finite float, greater than zero
- `children`: integer, greater than or equal to zero
- `smoker`: `yes` or `no`
- `region`: `northeast`, `northwest`, `southeast`, or `southwest`

Validation requirements:

- Reject unknown fields.
- Reject unsupported enum values.
- Reject non-finite numeric values.
- Use the same validation rules for the HTML form and JSON endpoint.
- Do not duplicate validation rules in route handlers or the service.

Create response schemas for:

- Prediction response:
  - `predicted_charges: float`
  - `currency: "USD"`
- Health response:
  - `status: "ok"`

Declare Pydantic as a direct project dependency.

### 2. Add an artifact repository layer

Introduce an `ArtifactRepository` protocol or abstract interface responsible for persisted model artifacts.

The local filesystem implementation must:

- Load `model.pkl`
- Load `preprocessor.pkl`
- Save the trained model
- Save the fitted preprocessor
- Accept artifact paths through constructor injection
- Raise application-specific exceptions for missing, unreadable, or invalid artifacts
- Contain all pickle/joblib filesystem persistence logic

Update the training components to save their artifacts through this repository instead of calling generic persistence helpers directly.

Do not add a database or prediction-history persistence as part of this issue.

### 3. Add a prediction service

Create a framework-independent `PredictionService`.

It must:

- Receive its artifact repository through constructor injection.
- Accept validated prediction input.
- Build the ordered feature input expected by the preprocessor:
  - `age`
  - `sex`
  - `bmi`
  - `children`
  - `smoker`
  - `region`
- Apply the persisted preprocessor.
- Run the persisted model.
- Return a normalized scalar prediction result.
- Raise explicit application exceptions instead of `HTTPException`.
- Avoid any imports from FastAPI.
- Avoid knowledge of HTTP responses or templates.
- Avoid any AWS or S3 behavior.

Load and cache the model artifacts once per service/application lifecycle rather than reading both pickle files for every request.

### 4. Make the API layer lean

Split the health, HTML, and JSON routes into API route modules.

Route handlers should only:

1. Receive and validate input.
2. Resolve the injected `PredictionService`.
3. Call the service.
4. Map expected application errors to HTTP or template responses.
5. Serialize the response.

Route handlers must not:

- Construct pandas DataFrames
- Know artifact paths
- Load pickle files
- Instantiate model infrastructure directly
- Contain prediction logic
- Catch an unrestricted `Exception` without logging and mapping it appropriately

Use a dependency provider in the application composition layer to construct:

- `LocalArtifactRepository`
- `PredictionService`

The dependency must be easy to override in FastAPI tests.

### 5. Preserve the API contract

Keep these endpoints:

- `GET /`
- `GET /health`
- `POST /predict`
- `POST /predict-json`

Expected behavior:

- `/health` returns `{"status": "ok"}`.
- `/predict` renders the existing HTML result.
- Invalid form data renders a safe validation message with HTTP 422.
- `/predict-json` returns a validated JSON response.
- Invalid JSON input returns HTTP 422.
- Unexpected inference errors return a sanitized HTTP 500 response.
- Missing or unreadable model artifacts return a sanitized HTTP 503 response.
- Internal paths, pickle errors, and stack traces are not exposed to clients.
- Existing request field names and categorical values remain unchanged.

### 6. Remove legacy prediction code

Remove or replace:

- `pipeline/predict_pipeline.py`
- `CustomData`
- `PredictPipeline`
- Direct `load_object` and `save_object` usage outside the repository layer
- Any obsolete persistence helpers left in `src/utils.py`
- Tests that monkeypatch internal artifact paths or S3 attributes

Do not leave compatibility wrappers that duplicate the new service and repository implementations.

### 7. Remove all AWS-specific material

Delete:

- `.ebextensions/`
- `deploy/ha/`
- `buildspec-train.yml`

Remove:

- `boto3` from `requirements.txt`
- `boto3` from `pyproject.toml`
- S3 download methods and imports
- `MODEL_S3_BUCKET`
- `MODEL_S3_MODEL_KEY`
- `MODEL_S3_PREPROCESSOR_KEY`
- S3-related test setup
- Elastic Beanstalk URLs
- Elastic Beanstalk deployment instructions
- AWS/EB references in `README.md`
- The EB reference in `docs/issue-mvp-demo.md`

Regenerate `uv.lock` so that AWS-only transitive dependencies such as `boto3`, `botocore`, and `s3transfer` are removed.

If `.env.example` becomes empty, delete it and remove the redundant `env_file` entry from `docker-compose.yml`.

Retain platform-neutral assets such as:

- `Dockerfile`
- `docker-compose.yml`
- GitHub Actions CI
- `/health`
- `Procfile`, provided it remains valid and contains no AWS-specific assumptions

### 8. Update entry points and documentation

Update all application entry points to the new import path, including:

- Docker `CMD`
- `Procfile`
- `Makefile`
- README commands
- Tests

Use a working command similar to:

```bash
uv run uvicorn medical_insurance.main:app --reload
```

Update packaging configuration for the `src` layout.

Update the README to include:

- The new architecture
- Local setup instructions
- How to train and create local artifacts
- How to run the API
- `localhost` curl examples
- The requirement that local artifacts exist before prediction
- No live AWS URL or AWS deployment instructions

Template and artifact paths must be resolved independently of the process working directory.

## Testing

Add or refactor tests for the following.

### Schema tests

- Valid input is accepted.
- Unknown fields are rejected.
- Invalid enum values are rejected.
- Negative age is rejected.
- Non-positive BMI is rejected.
- Negative children count is rejected.
- Non-finite numeric values are rejected.

### Repository tests

- Model and preprocessor artifacts can be saved and loaded from a temporary directory.
- Missing artifacts raise the expected repository exception.
- Corrupt artifacts raise a safe repository exception.

### Service tests

Use fake repository, preprocessor, and model objects to verify:

- Feature order is correct.
- Preprocessing is called.
- Prediction is called.
- A scalar result is returned.
- Artifacts are not loaded on every prediction.
- Repository errors are mapped to application-level service errors.

### API tests

Override the prediction service dependency instead of training a real model.

Verify:

- Health endpoint
- HTML endpoint
- Valid form prediction
- Invalid form prediction
- Valid JSON prediction
- Invalid JSON prediction
- Service failure response
- Missing-artifact response

Retain a separate integration test for the real training-to-prediction pipeline.

## Acceptance criteria

- [ ] FastAPI routes contain no pandas, scikit-learn, pickle, boto3, or artifact-path logic.
- [ ] Pydantic schemas are defined outside the route modules.
- [ ] HTML and JSON inputs use the same validation model.
- [ ] Prediction orchestration lives in a framework-independent service.
- [ ] Model persistence is accessed through an injected repository interface.
- [ ] Training and inference use the repository layer for model artifacts.
- [ ] Artifacts are not reloaded for every request.
- [ ] Existing endpoint paths and successful response formats remain compatible.
- [ ] Missing artifacts produce a controlled response rather than an exposed traceback.
- [ ] All AWS, S3, CodeBuild, Elastic Beanstalk, and `.ebextensions` code/configuration/documentation is removed.
- [ ] `boto3`, `botocore`, and `s3transfer` are absent from the regenerated dependency lock.
- [ ] Docker, Makefile, Procfile, README, and tests use the new application import path.
- [ ] `uv run --extra dev pytest` passes.
- [ ] `uv run --extra dev ruff check .` passes.
- [ ] A case-insensitive search, excluding `.git` and notebook binary content, finds no remaining AWS/S3/Elastic Beanstalk references.

## Out of scope

- Changing the regression algorithm or retraining strategy
- Changing the public request field names
- Adding authentication
- Adding a database
- Persisting prediction history
- Replacing AWS with another cloud provider
- Redesigning the HTML interface
